src/pybind/mgr/ssh/module.py

```diff
@@ -1378,4 +1376,101 @@ def _remove_rbd_mirror(daemons):
+
+
+class BaseScheduler(object):
+    """
+    Base Scheduler Interface
+
+    * requires a placement_spec
+
+    `place(host_pool)` needs to return a List[HostSpec, ..]
+    """
+
+    def __init__(self, placement_spec):
+        # type: (orchestrator.PlacementSpec) -> None
+        self.placement_spec = placement_spec
+
+    def place(self):
+        raise NotImplementedError
+
+
+class SimpleScheduler(BaseScheduler):
+    """
+    The most simple way to pick/schedule a set of hosts.
+    1) Shuffle the provided host_pool
+    2) Select from list up to :count
+    """
+    def __init__(self, placement_spec):
+        super(SimpleScheduler, self).__init__(placement_spec)
+
+    def place(self, host_pool, count=None):
+        # type: (List, Optional(int)) -> List
+        if not host_pool:
+            raise Exception('List of host candidates is empty')
+        host_pool = [HostSpec(x, '', '') for x in host_pool]
+        # shuffle for pseudo random selection
+        random.shuffle(host_pool)
+        return host_pool[:count]
+
+
+class NodeAssignment(object):
+    """
+    A class to detect if nodes are being passed imperative or declarative
+    If the spec is populated via the `nodes/hosts` field it will not load
+    any nodes into the list.
+    If the spec isn't populated, i.e. when only num or label is present (declarative)
+    it will use the provided `get_host_func` to load it from the inventory.
+
+    Schedulers can be assigned to pick hosts from the pool.
+    """
+
+    def __init__(self,
+                 spec=None,  # type: orchestrator.PlacementSpec
+                 scheduler=None,  # type: BaseScheduler
+                 get_hosts_func=None,  # type: Callable
+                 ):
+        self.spec = spec
+        self.scheduler = scheduler if scheduler else SimpleScheduler(self.spec.placement)
+        self.get_hosts_func = get_hosts_func
+
+    def load(self):
+        # type: () -> orchestrator.PlacementSpec
+        """
+        Load nodes into the spec.placement.nodes container.
+        """
+        self.load_labeled_nodes()
+        self.assign_nodes()
+        return self.spec
+
+    def load_labeled_nodes(self):
+        # type: () -> None
+        """
+        Assign nodes based on their label
+        """
+        # Querying for labeled nodes doesn't work currently.
+        # Leaving this open for the next iteration
+        # NOTE: This currently queries for all hosts without label restriction
+        if self.spec.placement.label:
+            logger.info("Found labels. Assinging nodes that match the label")
+            candidates = [HostSpec(x[0], '', '') for x in self.get_hosts_func()]  # TODO: query for labels
+            logger.info('Assigning nodes to spec: {}'.format(candidates))
+            self.spec.placement.set_nodes(candidates)
+
+    def assign_nodes(self):
+        # type: () -> None
+        """
+        Use the assigned scheduler to load nodes into the spec.placement.nodes container
```

Comment:
Would it make more sense to break this down differently:
- First pick nodes.  If no label, all nodes.  If label, all labeled nodes.  If node list specified, that list.
- Schedule services across those nodes.

Right now the scheduler is only used if no label is used. But I would think that if you say "2 RGWs" but have 5 nodes labeled you still want only 2 RGWs (but if no count is specified you want 5).  Or, if you say 2 RGWs and pass a list of 3 nodes (don't ask me why) we'd want the pick the 2 best of that set.  Or is that too weird?

Reply:
So currently it's like this:

* If a node list is provided, use that list. No scheduling/placement algorithm is necessary.
* If a label is provided, use the hosts that have this label (currently not limited by labels, due missing label filters)
* If num/count is provided, use all hosts and let the scheduler decide where to place the service.
* If no label and no node list and no num/count, We probably can't draw any conclusion (not sure if that's the right tactic, though)

so I think it's already doing what you've asked for. Let's have a look at an example.

You have 5 hosts (node{1..5}, node{1,2} are labeld with rgw.

If you now want to have two RGWs you have a couple of options.

1) `ceph orchestrator update zone_ex realm_ex 2 node1, node2`
node1 and node2 are used. (no scheduler/placement as we specified a node list)

2) `ceph orchestrator update zone_ex realm_ex --label rgw`
node1 and node2 are used (loaded from the available host_list with label: rgw)
Also, we don't need the scheduler, as count we want to deploy on all nodes with label 'rgw'.

Now the special cases: We want to use only *one* node but don't really care where this goes.

3) `ceph orchestrator update zone_ex ream_ex 2`
two of node{1..5} are used (chosen by the scheduler.)

We want to use *one* node but it should have the label 'rgw'

4) `ceph orchestrator update zone_ex ream_ex 2 --label rgw`
*one* node of node{1..2} is used (chosen by the scheduler.)

All these cases should work with the current implementation. Is that enough, or are there more cases?

Reply:
5. Count is specified, and nodes are labeled by type.  In this case, we want to deploy, say, 1 MDS for a new fs, but we have 8 nodes with label ``mds``.  My goal here is to have services set up without *any* extra user args (e.g., default of 1 service or whatever when a new fs volume is created) and have the mds service appropriately scheduled without further thought.  

I'm not actually sure if this is what rook does: if you specify a node label and a count that is smaller than the number of labeled nodes, will it schedule among them?

In any case, what I'm thinking is that it'd be nice if the user just did some simple node labels one time (these nodes are for monitors, these are for gateways and mdss, these are for osds) and the rest of the time they don't need to do anything special--not even specifying the label names, since they have a default name.

Reply:
So, allow a bit more magic behind the scenes. I think I do like the 'label once and don't care about providing the --label flags anymore' type thing. I just guess that we still have to keep the explicit --label option. I'll have a look into that.

Reply:
done via https://github.com/ceph/ceph/pull/32003/commits/bc9428248d71b7b9f71773a60228081ddfd2d7fc